File: 125 Valid Palindrome/125 Valid Palindrome.swift
class Solution {
  func isPalindrome(_ s: String) -> Bool {
    let characters: [Character] = Array(s)
    var left: Int = 0
    var right: Int = characters.count - 1

    while left < right {
      while left < right && !characters[left].isAlphanumeric() {
        left += 1
      }

      while left < right && !characters[right].isAlphanumeric() {
        right -= 1
      }

      guard
        characters[left].lowercased() == characters[right].lowercased()
      else {
        return false
      }

      left += 1
      right -= 1
    }

    return true
  }
}

extension Character {
  func isAlphanumeric() -> Bool {
    return self.isLetter || self.isNumber
  }
}

File: 125 Valid Palindrome/125 Valid Palindrome.ts
function isPalindrome(s: string): boolean {
  let left: number = 0;
  let right: number = s.length - 1;

  while (left < right) {
    while (left < right && !isAlphanumeric(s[left])) {
      left++;
    }

    while (left < right && !isAlphanumeric(s[right])) {
      right--;
    }

    if (s[left].toLowerCase() !== s[right].toLowerCase()) {
      return false;
    }

    left++;
    right--;
  }

  return true;
}

function isAlphanumeric(s: string): boolean {
  const charCode: number = s.charCodeAt(0);

  return (
    (charCode >= 'A'.charCodeAt(0) && charCode <= 'Z'.charCodeAt(0)) ||
    (charCode >= 'a'.charCodeAt(0) && charCode <= 'z'.charCodeAt(0)) ||
    (charCode >= '0'.charCodeAt(0) && charCode <= '9'.charCodeAt(0))
  );
}
